### PR TITLE
Service retries and handling rate limit (429) and forbidden error (403)

### PR DESCRIPTION
**Pull request by johnwu96822**
The service error handling should match sand-ruby now.

Also change to just one global cache instead of several with different default expiration time. The default expiration time of a cache actually shouldn't matter to us, since we always specify the expiration duration in every cache Set call.

- [ ] @johnny-lai 
- [x] @abdulchaudhrycoupa 
- [ ] @supriya

## Generated Reviewers

<details>
  <summary><em>:warning: Do not modify anything in this section! :warning:</em></summary>
  <em>The content in this section is managed automatically, and any manual changes to it will be ignored and overwritten the next time the code review status is refreshed! View <a href="https://cody.coupa.engineering/repos/coupa/sand-go/pull/22">this PR</a> on Cody to see the current code review status.</em>
</details>

### CODE REVIEW

**Comment by johnwu96822** [comment]
```
AMER-M-611:sand-go johnwu$ go test ./... -cover
ok  	github.com/coupa/sand-go	15.597s	coverage: 91.7% of statements
ok  	github.com/coupa/sand-go/cache	0.315s	coverage: 100.0% of statements
AMER-M-611:sand-go johnwu$ 
```

File: service.go [hunk 1]
```diff
@@ -219,44 +222,66 @@ func (s *Service) verifyToken(token string, opt *VerificationOption) (map[string
 	dBytes, _ := json.Marshal(data)
 	req, _ := http.NewRequest("POST", s.TokenVerifyURL, bytes.NewBuffer(dBytes))
 	req.Header.Add("Authorization", "Bearer "+accessToken)
+
+	retryCount := 0
+
+RETRY:
```
[the remark this answers is not in the document]
Comment (johnwu96822) [reply service.go]: Changed it to a loop.

File: service.go [hunk 2]
```diff
@@ -219,44 +223,62 @@ func (s *Service) verifyToken(token string, opt *VerificationOption) (map[string
 	dBytes, _ := json.Marshal(data)
```
Comment (abdulchaudhrycoupa) [inline service.go:223]: @johnwu96822 don't you want to handle this error, as in bad JSON?

File: service.go [hunk 3]
```diff
@@ -219,44 +223,62 @@ func (s *Service) verifyToken(token string, opt *VerificationOption) (map[string
...
 	req, _ := http.NewRequest("POST", s.TokenVerifyURL, bytes.NewBuffer(dBytes))
```
Comment (abdulchaudhrycoupa) [inline service.go:224]: @johnwu96822 again, don't you want to handle this error?

**Review by abdulchaudhrycoupa — APPROVED** [review]

File: service.go [hunk 4]
```diff
@@ -216,47 +220,79 @@ func (s *Service) verifyToken(token string, opt *VerificationOption) (map[string
...
+		req, err := http.NewRequest("POST", s.TokenVerifyURL, bytes.NewBuffer(dBytes))
+		if err != nil {
+			return nil, fmt.Errorf("(%s) Service failed to create token verification request: %v", s.ClientID, err)
+		}
+		req.Header.Add("Authorization", "Bearer "+accessToken)
+		resp, err := client.Do(req)
+		if err != nil {
+			if retryCount < *opt.NumRetry {
+				sleep := math.Pow(2, float64(retryCount))
+				time.Sleep(time.Duration(sleep) * time.Second)
```
Comment (johnny-lai) [inline service.go:239]: You purposely changed the sleep in 
https://github.com/coupa/sand-go/pull/22/files#diff-0ddbbdfef22c6ffc6b6cc34938959181a43be0a70059d47bc430570baca3da8bR268-R270

Does this change need to be here? Perhaps you should add a function so it behaves consistently?

File: service.go [hunk 5]
```diff
@@ -216,47 +220,79 @@ func (s *Service) verifyToken(token string, opt *VerificationOption) (map[string
...
+					fmt.Sprintf("(%s) Service cannot access the token verification endpoint at this time with the current service token", s.ClientID),
+				}}
+			case http.StatusForbidden:
+				return nil, PermissionDeniedError{SandError{
+					fmt.Sprintf("(%s) Service is denied access to the token verification endpoint", s.ClientID),
+				}}
+			default:
+				if retryCount < *opt.NumRetry {
+					sleep := math.Pow(2, float64(retryCount))
+					time.Sleep(time.Duration(sleep) * time.Second)
```
Comment (johnny-lai) [inline service.go:267]: Use a function

File: service.go [hunk 6]
```diff
@@ -216,47 +220,79 @@ func (s *Service) verifyToken(token string, opt *VerificationOption) (map[string
...
+		defer resp.Body.Close()
+		body, _ = ioutil.ReadAll(resp.Body)
+		if resp.StatusCode != http.StatusOK {
+			str := fmt.Sprintf("(%s) Error response from the authentication service: %d - %s", s.ClientID, resp.StatusCode, body)
+			log.Warn(str)
+			switch resp.StatusCode {
+			case http.StatusInternalServerError:
+				//When the response is 500, the client's token may be expired. So let the client retry
+				//and return 401 by returning nil, so that the result is not cached.
+				return nil, nil
```
Comment (johnny-lai) [inline service.go:255]: Where does the retry occur?

Comment (johnny-lai) [reply service.go:255]: Also consider updating the comments for `VerifyTokenWithCache` to specify what `nil` for the map signifies?

File: service.go [hunk 7]
```diff
@@ -36,9 +37,9 @@ type Service struct {
 	TokenVerifyURL string
 
 	//The default expiry time for cache for invalid tokens and also valid tokens without expiry times
-	//Default value is 3600 (1 hour)
+	//Default value is 3599 seconds
 	//Only services need this because client tokens will always give expiry time
-	DefaultExpTime int
+	DefaultExpTime time.Duration
```
Comment (johnny-lai) [inline service.go:42]: Maybe `DefaultExpDuration` would be clearer?

File: service.go (hunk 6 above)
Comment (johnwu96822) [reply service.go:255]: Retry occurs on the client when it receives 401 response

**Review by supriya — APPROVED** [review]

**Comment by johnwu96822** [comment]
```
AMER-M-611:sand-go johnwu$ go test ./... -cover
ok  	github.com/coupa/sand-go	15.384s	coverage: 91.3% of statements
ok  	github.com/coupa/sand-go/cache	(cached)	coverage: 100.0% of statements
AMER-M-611:sand-go johnwu$
```

File: service.go (hunk 6 above)
Comment (johnny-lai) [reply service.go:242]: Oh I see. We are treating this as an invalid token to ask the client to fetch a new token.

**Review by johnny-lai — APPROVED** [review]

**Comment by johnwu96822** [comment]
```
AMER-M-611:sand-go johnwu$ go test ./... -cover
ok  	github.com/coupa/sand-go	15.311s	coverage: 91.4% of statements
ok  	github.com/coupa/sand-go/cache	0.235s	coverage: 100.0% of statements
AMER-M-611:sand-go johnwu$
```

**Comment by johnwu96822** [comment]
[lockit v4.9.3] merged via Lockit CLI with --skip-pending-status on 2021-09-06 16:04:20.4283 +0800 CST m=+3.894638589

**Pull request merged**